Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,560 @@
+//         ".."
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     }
+//     // `.Captures == (Substring, Substring)`
+```
+
+#### Named Capturing Group
+
+A named capturing group in a pattern with multiple captures causes the resulting
+tuple to have a tuple element label at the corresponding capture type position.
+When the pattern has only one capture, there will be no tuple element label
+because there are no 1-element tuples.
+
+```
+pattern ::= '(' '?' '<' name '>' pattern ')'
+```
+
+```swift
+let graphemeBreakLowerBound = /(?<lower>[0-9A-F]+)\.\.[0-9A-F]+/
+// => `Regex<Substring>`
+
+let graphemeBreakRange = /(?<lower>[0-9A-F]+)\.\.(?<upper>[0-9A-F]+)/
+// => `Regex<(lower: Substring, upper: Substring)>`
+```
+
+#### Non-capturing Group
+
+A non-capturing group's capture type is the capture type of its underlying pattern.
+
+```
+pattern ::= '(' '?' ':' pattern ')'
+```
+
+```swift
+let graphemeBreakLowerBound = /([0-9A-F]+)(?:\.\.([0-9A-F]+))?/
+// => `Regex<(Substring, Substring?)>`
+
+// Result builder DSL equivalent:
+//     let graphemeBreakLowerBound = Regex {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//         Optionally {
+//             ".."
+//             OneOrMore(CharacterClass.hexDigit).capture()
+//         }
+//     }
+//     // `.Captures == (Substring, Substring?)`
+```
+
+#### Nested capturing group
+
+When capturing group is nested within another capturing group, they count as two
+distinct captures in the order their left parenthesis first appears in the
+regular expression literal. This is consistent with PCRE and allows us to use
+backreferences (e.g. `\2`) with linear indices.
+
+```swift
+let graphemeBreakPropertyData = /(([0-9A-F]+)(\.\.([0-9A-F]+)))\s*;\s(\w+).*/
+// Positions in capture tuple: 0 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    4 ^~~~~
+//                                         2 ^~~~~~~~~~~~~~~~~
+//                              1 ^~~~~~~~~~~   3 ^~~~~~~~~~~
+// => `Regex<(Substring, Substring, Substring, Substring, Substring)>`
+
+// Result builder DSL equivalent:
+//     let graphemeBreakPropertyData = Regex {
+//         Regex {
+//             OneOrMore(CharacterClass.hexDigit).capture() // (1)
+//             Regex {
+//                 ".."
+//                 OneOrMore(CharacterClass.hexDigit).capture() // (3)
+//             }.capture() // (2)
+//         }.capture() // (0)
+//         Repeat(CharacterClass.whitespace)
+//         ";"
+//         CharacterClass.whitespace
+//         OneOrMore(CharacterClass.word).capture() // (4)
+//         Repeat(CharacterClass.any)
+//     }
+//     // `.Captures == Regex(Substring, Substring, Substring, Substring, Substring)`
+
+let input = "007F..009F    ; Control # Cc  [33] <control-007F>..<control-009F>"
+// Match result for `input`:
+// ("007F..009F", "007F", "..009F", "009F", "Control")
+```
+
+
+#### Quantification
+
+A quantifier's capture type depends on its underlying pattern.
```

Comment:
Should we note that this is a big difference from most implementations, where you only have access to the last match of a repeated capture group?

Reply:
For example, Python only gives you access to the last group of this dash-separated string:

```python
rep = re.compile('(?:([0-9a-f]+)-?)+')
match = rep.match("1234-5678-9abc-def0")
print(match.group(1))
# Prints "def0"
```

Whereas the proposed version will provide access to all four sub-matches:

```swift
// Note: Syntax / API still in discussion
let pattern = /(?:([0-9a-f]+)-?)+/
if let match = "1234-5678-9abc-def0".firstMatch(of: pattern) {
    print(match.captures)
}
// Prints ["1234", "5678", "9abc", "def0"]
```